File: packages/protocol/typescript/validators.ts
import type {
  CommandDirective,
  EditAction,
  EditDirective,
  PingResult,
  VoiceTranscriptCompletion,
  VoiceTranscriptDirective,
} from "./types.generated";

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function hasOnlyKeys(
  value: Record<string, unknown>,
  allowedKeys: string[],
): boolean {
  const allowed = new Set(allowedKeys);
  return Object.keys(value).every((key) => allowed.has(key));
}

function isAnchor(value: unknown): value is { before: string; after: string } {
  return (
    isRecord(value) &&
    hasOnlyKeys(value, ["before", "after"]) &&
    typeof value.before === "string" &&
    typeof value.after === "string"
  );
}

export function isPingResult(value: unknown): value is PingResult {
  return isRecord(value) && value.message === "pong";
}

export function isEditAction(value: unknown): value is EditAction {
  const isReplaceBetweenAnchors =
    isRecord(value) &&
    hasOnlyKeys(value, ["kind", "path", "anchor", "newText", "editId"]) &&
    value.kind === "replace_between_anchors" &&
    typeof value.path === "string" &&
    isAnchor(value.anchor) &&
    typeof value.newText === "string" &&
    (value.editId === undefined || typeof value.editId === "string");
  if (isReplaceBetweenAnchors) return true;

  const isCreateFile =
    isRecord(value) &&
    hasOnlyKeys(value, ["kind", "path", "content", "editId"]) &&
    value.kind === "create_file" &&
    typeof value.path === "string" &&
    typeof value.content === "string" &&
    (value.editId === undefined || typeof value.editId === "string");
  if (isCreateFile) return true;

  return (
    isRecord(value) &&
    hasOnlyKeys(value, ["kind", "path", "text", "editId"]) &&
    value.kind === "append_to_file" &&
    typeof value.path === "string" &&
    typeof value.text === "string" &&
    (value.editId === undefined || typeof value.editId === "string")
  );
}

export function isEditDirective(value: unknown): value is EditDirective {
  if (!isRecord(value) || typeof value.kind !== "string") {
    return false;
  }

  switch (value.kind) {
    case "success":
      return (
        hasOnlyKeys(value, ["kind", "actions"]) &&
        Array.isArray(value.actions) &&
        value.actions.every(isEditAction)
      );
    case "noop":
      return (
        hasOnlyKeys(value, ["kind", "reason"]) &&
        typeof value.reason === "string"
      );
    default:
      return false;
  }
}

function isCommandDirective(value: unknown): value is CommandDirective {
  if (!isRecord(value)) {
    return false;
  }

  if (
    !hasOnlyKeys(value as Record<string, unknown>, [
      "command",
      "args",
      "timeoutMs",
    ])
  ) {
    return false;
  }

  if (typeof (value as Record<string, unknown>).command !== "string") {
    return false;
  }

  const args = (value as Record<string, unknown>).args;
  if (args !== undefined) {
    if (!Array.isArray(args) || !args.every((x) => typeof x === "string")) {
      return false;
    }
  }

  const timeoutMs = (value as Record<string, unknown>).timeoutMs;
  if (timeoutMs !== undefined && typeof timeoutMs !== "number") {
    return false;
  }

  return true;
}

function isNavigationAction(value: unknown): boolean {
  if (!isRecord(value) || typeof value.kind !== "string") return false;
  switch (value.kind) {
    case "open_file":
      return (
        hasOnlyKeys(value, ["kind", "openFile"]) &&
        isRecord(value.openFile) &&
        hasOnlyKeys(value.openFile, ["path"]) &&
        typeof value.openFile.path === "string"
      );
    case "reveal_symbol":
      return (
        hasOnlyKeys(value, ["kind", "revealSymbol"]) &&
        isRecord(value.revealSymbol) &&
        hasOnlyKeys(value.revealSymbol, ["path", "symbolName", "symbolKind"]) &&
        typeof value.revealSymbol.symbolName === "string" &&
        (value.revealSymbol.path === undefined ||
          typeof value.revealSymbol.path === "string") &&
        (value.revealSymbol.symbolKind === undefined ||
          typeof value.revealSymbol.symbolKind === "string")
      );
    case "move_cursor":
      return (
        hasOnlyKeys(value, ["kind", "moveCursor"]) &&
        isRecord(value.moveCursor) &&
        hasOnlyKeys(value.moveCursor, ["target"]) &&
        isRecord(value.moveCursor.target) &&
        hasOnlyKeys(value.moveCursor.target, ["path", "line", "char"]) &&
        typeof value.moveCursor.target.line === "number" &&
        typeof value.moveCursor.target.char === "number" &&
        (value.moveCursor.target.path === undefined ||
          typeof value.moveCursor.target.path === "string")
      );
    case "select_range":
      return (
        hasOnlyKeys(value, ["kind", "selectRange"]) &&
        isRecord(value.selectRange) &&
        hasOnlyKeys(value.selectRange, ["target"]) &&
        isRecord(value.selectRange.target) &&
        hasOnlyKeys(value.selectRange.target, [
          "path",
          "startLine",
          "startChar",
          "endLine",
          "endChar",
        ]) &&
        typeof value.selectRange.target.startLine === "number" &&
        typeof value.selectRange.target.startChar === "number" &&
        typeof value.selectRange.target.endLine === "number" &&
        typeof value.selectRange.target.endChar === "number" &&
        (value.selectRange.target.path === undefined ||
          typeof value.selectRange.target.path === "string")
      );
    case "reveal_edit":
      return (
        hasOnlyKeys(value, ["kind", "revealEdit"]) &&
        isRecord(value.revealEdit) &&
        hasOnlyKeys(value.revealEdit, ["editId"]) &&
        typeof value.revealEdit.editId === "string"
      );
    default:
      return false;
  }
}

function isNavigationDirective(value: unknown): boolean {
  if (!isRecord(value) || typeof value.kind !== "string") {
    return false;
  }
  if (value.kind === "success") {
    return (
      hasOnlyKeys(value, ["kind", "action"]) && isNavigationAction(value.action)
    );
  }
  if (value.kind === "noop") {
    return (
      hasOnlyKeys(value, ["kind", "reason"]) && typeof value.reason === "string"
    );
  }
  return false;
}

function isUndoDirective(value: unknown): boolean {
  return (
    isRecord(value) &&
    hasOnlyKeys(value, ["scope"]) &&
    typeof value.scope === "string" &&
    (value.scope === "last_edit" || value.scope === "last_transcript")
  );
}

export function isVoiceTranscriptDirective(
  value: unknown,
): value is VoiceTranscriptDirective {
  if (!isRecord(value) || typeof value.kind !== "string") {
    return false;
  }
  if (value.kind === "edit") {
    return (
      hasOnlyKeys(value, ["kind", "editDirective"]) &&
      isEditDirective(value.editDirective)
    );
  }
  if (value.kind === "command") {
    return (
      hasOnlyKeys(value, ["kind", "commandDirective"]) &&
      isCommandDirective(value.commandDirective)
    );
  }
  if (value.kind === "navigate") {
    return (
      hasOnlyKeys(value, ["kind", "navigationDirective"]) &&
      isNavigationDirective(value.navigationDirective)
    );
  }
  if (value.kind === "undo") {
    return (
      hasOnlyKeys(value, ["kind", "undoDirective"]) &&
      isUndoDirective(value.undoDirective)
    );
  }
  if (value.kind === "rename") {
    return (
      hasOnlyKeys(value, ["kind", "renameDirective"]) &&
      isRenameDirective(value.renameDirective)
    );
  }
  if (value.kind === "code_action") {
    return (
      hasOnlyKeys(value, ["kind", "codeActionDirective"]) &&
      isCodeActionDirective(value.codeActionDirective)
    );
  }
  if (value.kind === "format") {
    return (
      hasOnlyKeys(value, ["kind", "formatDirective"]) &&
      isFormatDirective(value.formatDirective)
    );
  }
  if (value.kind === "delete_file") {
    return (
      hasOnlyKeys(value, ["kind", "deleteFileDirective"]) &&
      isDeleteFileDirective(value.deleteFileDirective)
    );
  }
  if (value.kind === "move_path") {
    return (
      hasOnlyKeys(value, ["kind", "movePathDirective"]) &&
      isMovePathDirective(value.movePathDirective)
    );
  }
  if (value.kind === "create_folder") {
    return (
      hasOnlyKeys(value, ["kind", "createFolderDirective"]) &&
      isCreateFolderDirective(value.createFolderDirective)
    );
  }
  return false;
}

function isRenameDirective(value: unknown): boolean {
  return (
    isRecord(value) &&
    hasOnlyKeys(value, ["path", "position", "newName"]) &&
    typeof value.path === "string" &&
    value.path.trim() !== "" &&
    typeof value.newName === "string" &&
    value.newName.trim() !== "" &&
    isRecord(value.position) &&
    hasOnlyKeys(value.position as Record<string, unknown>, [
      "line",
      "character",
    ]) &&
    typeof (value.position as { line: unknown }).line === "number" &&
    Number.isInteger((value.position as { line: number }).line) &&
    (value.position as { line: number }).line >= 0 &&
    typeof (value.position as { character: unknown }).character === "number" &&
    Number.isInteger((value.position as { character: number }).character) &&
    (value.position as { character: number }).character >= 0
  );
}

const codeActionKinds = new Set([
  "refactor.extract.function",
  "refactor.extract.variable",
  "refactor.extract.constant",
  "refactor.inline",
  "source.organizeImports",
  "source.fixAll",
  "quickfix",
]);

function isCodeActionDirective(value: unknown): boolean {
  if (!isRecord(value)) return false;
  const keys = [
    "path",
    "actionKind",
    "range",
    "preferredTitleIncludes",
  ] as const;
  if (!hasOnlyKeys(value, [...keys])) return false;
  if (typeof value.path !== "string" || value.path.trim() === "") return false;
  if (
    typeof value.actionKind !== "string" ||
    !codeActionKinds.has(value.actionKind)
  ) {
    return false;
  }
  if (value.range !== undefined) {
    if (!isRecord(value.range)) return false;
    const r = value.range as Record<string, unknown>;
    if (
      !hasOnlyKeys(r, ["startLine", "startChar", "endLine", "endChar"]) ||
      typeof r.startLine !== "number" ||
      !Number.isInteger(r.startLine) ||
      r.startLine < 0 ||
      typeof r.startChar !== "number" ||
      !Number.isInteger(r.startChar) ||
      r.startChar < 0 ||
      typeof r.endLine !== "number" ||
      !Number.isInteger(r.endLine) ||
      r.endLine < 0 ||
      typeof r.endChar !== "number" ||
      !Number.isInteger(r.endChar) ||
      r.endChar < 0
    ) {
      return false;
    }
  }
  if (
    value.preferredTitleIncludes !== undefined &&
    typeof value.preferredTitleIncludes !== "string"
  ) {
    return false;
  }
  return true;
}

function isFormatDirective(value: unknown): boolean {
  if (!isRecord(value)) return false;
  if (!hasOnlyKeys(value, ["path", "scope", "range"])) return false;
  if (typeof value.path !== "string" || value.path.trim() === "") return false;
  if (value.scope !== "document" && value.scope !== "selection") return false;
  if (value.range !== undefined) {
    if (!isRecord(value.range)) return false;
    const r = value.range as Record<string, unknown>;
    if (
      !hasOnlyKeys(r, ["startLine", "startChar", "endLine", "endChar"]) ||
      typeof r.startLine !== "number" ||
      !Number.isInteger(r.startLine) ||
      r.startLine < 0 ||
      typeof r.startChar !== "number" ||
      !Number.isInteger(r.startChar) ||
      r.startChar < 0 ||
      typeof r.endLine !== "number" ||
      !Number.isInteger(r.endLine) ||
      r.endLine < 0 ||
      typeof r.endChar !== "number" ||
      !Number.isInteger(r.endChar) ||
      r.endChar < 0
    ) {
      return false;
    }
  }
  return true;
}

function isDeleteFileDirective(value: unknown): boolean {
  return (
    isRecord(value) &&
    hasOnlyKeys(value, ["path"]) &&
    typeof value.path === "string" &&
    value.path.trim() !== ""
  );
}

function isMovePathDirective(value: unknown): boolean {
  return (
    isRecord(value) &&
    hasOnlyKeys(value, ["from", "to"]) &&
    typeof value.from === "string" &&
    value.from.trim() !== "" &&
    typeof value.to === "string" &&
    value.to.trim() !== ""
  );
}

function isCreateFolderDirective(value: unknown): boolean {
  return (
    isRecord(value) &&
    hasOnlyKeys(value, ["path"]) &&
    typeof value.path === "string" &&
    value.path.trim() !== ""
  );
}

// biome-ignore lint/complexity/noExcessiveCognitiveComplexity: Validator is expected to be complex (exhaustive)
export function isVoiceTranscriptCompletion(
  value: unknown,
): value is VoiceTranscriptCompletion {
  if (!isRecord(value) || typeof value.success !== "boolean") {
    return false;
  }
  const allowedKeys = new Set([
    "success",
    "summary",
    "transcriptOutcome",
    "uiDisposition",
    "searchResults",
    "activeSearchIndex",
    "answerText",
    "clarifyTargetResolution",
    "fileSelectionFocusPath",
  ]);
  if (!Object.keys(value).every((k) => allowedKeys.has(k))) {
    return false;
  }
  if (value.summary !== undefined) {
    if (typeof value.summary !== "string") {
      return false;
    }
    if ([...value.summary].length > 8192) {
      return false;
    }
  }
  if (value.success !== true && value.summary !== undefined) {
    return false;
  }
  if (value.transcriptOutcome !== undefined) {
    if (value.success !== true) {
      return false;
    }
    if (
      value.transcriptOutcome !== "irrelevant" &&
      value.transcriptOutcome !== "completed" &&
      value.transcriptOutcome !== "clarify" &&
      value.transcriptOutcome !== "clarify_control" &&
      value.transcriptOutcome !== "search" &&
      value.transcriptOutcome !== "selection" &&
      value.transcriptOutcome !== "selection_control" &&
      value.transcriptOutcome !== "file_selection" &&
      value.transcriptOutcome !== "file_selection_control" &&
      value.transcriptOutcome !== "needs_workspace_folder" &&
      value.transcriptOutcome !== "answer"
    ) {
      return false;
    }
  }

  if (value.clarifyTargetResolution !== undefined) {
    if (value.success !== true) {
      return false;
    }
    if (typeof value.clarifyTargetResolution !== "string") {
      return false;
    }
  }

  if (value.uiDisposition !== undefined) {
    if (value.success !== true) {
      return false;
    }
    if (
      value.uiDisposition !== "shown" &&
      value.uiDisposition !== "skipped" &&
      value.uiDisposition !== "hidden"
    ) {
      return false;
    }
  }

  if (value.searchResults !== undefined) {
    if (value.success !== true) {
      return false;
    }
    if (!Array.isArray(value.searchResults)) {
      return false;
    }
    for (const item of value.searchResults) {
      if (!isRecord(item)) {
        return false;
      }
      if (!hasOnlyKeys(item, ["path", "line", "character", "preview"])) {
        return false;
      }
      const rec = item as Record<string, unknown>;
      if (typeof rec.path !== "string") {
        return false;
      }
      if (!Number.isInteger(rec.line) || (rec.line as number) < 0) {
        return false;
      }
      if (!Number.isInteger(rec.character) || (rec.character as number) < 0) {
        return false;
      }
      if (typeof rec.preview !== "string") {
        return false;
      }
    }
  }
  if (value.activeSearchIndex !== undefined) {
    if (value.success !== true) {
      return false;
    }
    if (
      typeof value.activeSearchIndex !== "number" ||
      !Number.isInteger(value.activeSearchIndex) ||
      value.activeSearchIndex < 0
    ) {
      return false;
    }
  }
  if (value.answerText !== undefined) {
    if (value.success !== true) {
      return false;
    }
    if (typeof value.answerText !== "string") {
      return false;
    }
    if ([...value.answerText].length > 8192) {
      return false;
    }
  }
  if (value.fileSelectionFocusPath !== undefined) {
    if (value.success !== true) {
      return false;
    }
    if (typeof value.fileSelectionFocusPath !== "string") {
      return false;
    }
  }
  if (value.transcriptOutcome === "file_selection_control") {
    if (
      typeof value.fileSelectionFocusPath !== "string" ||
      value.fileSelectionFocusPath.trim() === ""
    ) {
      return false;
    }
  }
  return true;
}

File: packages/protocol/go/validators.go
package protocol

import (
	"errors"
	"fmt"
	"strings"
)

// EditDirective validation lives here alongside future protocol-level validators
// (mirrors typescript/validators.ts conceptually).

func (r EditDirective) Validate() error {
	switch r.Kind {
	case "success":
		if r.Actions == nil {
			return errors.New("success result must include actions")
		}
		if r.Reason != "" {
			return errors.New("success result must not contain reason")
		}
	case "noop":
		if r.Reason == "" {
			return errors.New("noop result must include reason")
		}
		if len(r.Actions) > 0 {
			return errors.New("noop result must not contain actions")
		}
	default:
		return errors.New("unknown edit.dispatch result kind")
	}

	return nil
}

func (s VoiceTranscriptDirective) Validate() error {
	switch s.Kind {
	case "edit":
		if s.EditDirective == nil || s.CommandDirective != nil || s.NavigationDirective != nil || s.UndoDirective != nil ||
			s.RenameDirective != nil || s.CodeActionDirective != nil || s.FormatDirective != nil ||
			s.DeleteFileDirective != nil || s.MovePathDirective != nil || s.CreateFolderDirective != nil {
			return errors.New("voice transcript step: kind edit requires editDirective and no other directive payloads")
		}
		return s.EditDirective.Validate()
	case "command":
		if s.CommandDirective == nil || s.EditDirective != nil || s.NavigationDirective != nil || s.UndoDirective != nil ||
			s.RenameDirective != nil || s.CodeActionDirective != nil || s.FormatDirective != nil ||
			s.DeleteFileDirective != nil || s.MovePathDirective != nil || s.CreateFolderDirective != nil {
			return errors.New("voice transcript step: kind command requires commandDirective and no other directive payloads")
		}
		if strings.TrimSpace(s.CommandDirective.Command) == "" {
			return errors.New("voice transcript step: command requires non-empty commandDirective.command")
		}
		// CommandDirective has no additional protocol-level validation yet; host-side
		// policy executes the safety checks.
		return nil
	case "navigate":
		if s.NavigationDirective == nil || s.EditDirective != nil || s.CommandDirective != nil || s.UndoDirective != nil ||
			s.RenameDirective != nil || s.CodeActionDirective != nil || s.FormatDirective != nil ||
			s.DeleteFileDirective != nil || s.MovePathDirective != nil || s.CreateFolderDirective != nil {
			return errors.New("voice transcript step: kind navigate requires navigationDirective and no other directive payloads")
		}
		return validateNavigationDirective(s.NavigationDirective)
	case "undo":
		if s.UndoDirective == nil || s.EditDirective != nil || s.CommandDirective != nil || s.NavigationDirective != nil ||
			s.RenameDirective != nil || s.CodeActionDirective != nil || s.FormatDirective != nil ||
			s.DeleteFileDirective != nil || s.MovePathDirective != nil || s.CreateFolderDirective != nil {
			return errors.New("voice transcript step: kind undo requires undoDirective and no other directive payloads")
		}
		return validateUndoDirective(s.UndoDirective)
	case "rename":
		if s.RenameDirective == nil || s.EditDirective != nil || s.CommandDirective != nil || s.NavigationDirective != nil || s.UndoDirective != nil ||
			s.CodeActionDirective != nil || s.FormatDirective != nil ||
			s.DeleteFileDirective != nil || s.MovePathDirective != nil || s.CreateFolderDirective != nil {
			return errors.New("voice transcript step: kind rename requires renameDirective and no other directive payloads")
		}
		r := s.RenameDirective
		if strings.TrimSpace(r.Path) == "" || strings.TrimSpace(r.NewName) == "" {
			return errors.New("voice transcript step: rename requires path and newName")
		}
		if r.Position.Line < 0 || r.Position.Character < 0 {
			return errors.New("voice transcript step: rename requires non-negative position")
		}
		return nil
	case "code_action":
		if s.CodeActionDirective == nil || s.EditDirective != nil || s.CommandDirective != nil || s.NavigationDirective != nil || s.UndoDirective != nil ||
			s.RenameDirective != nil || s.FormatDirective != nil ||
			s.DeleteFileDirective != nil || s.MovePathDirective != nil || s.CreateFolderDirective != nil {
			return errors.New("voice transcript step: kind code_action requires codeActionDirective and no other directive payloads")
		}
		c := s.CodeActionDirective
		if strings.TrimSpace(c.Path) == "" || strings.TrimSpace(c.ActionKind) == "" {
			return errors.New("voice transcript step: code_action requires path and actionKind")
		}
		if c.Range != nil {
			if c.Range.StartLine < 0 || c.Range.StartChar < 0 || c.Range.EndLine < 0 || c.Range.EndChar < 0 {
				return errors.New("voice transcript step: code_action range positions must be non-negative")
			}
		}
		return nil
	case "format":
		if s.FormatDirective == nil || s.EditDirective != nil || s.CommandDirective != nil || s.NavigationDirective != nil || s.UndoDirective != nil ||
			s.RenameDirective != nil || s.CodeActionDirective != nil ||
			s.DeleteFileDirective != nil || s.MovePathDirective != nil || s.CreateFolderDirective != nil {
			return errors.New("voice transcript step: kind format requires formatDirective and no other directive payloads")
		}
		f := s.FormatDirective
		if strings.TrimSpace(f.Path) == "" {
			return errors.New("voice transcript step: format requires path")
		}
		scope := strings.TrimSpace(f.Scope)
		if scope != "document" && scope != "selection" {
			return fmt.Errorf("voice transcript step: format scope must be document or selection, got %q", f.Scope)
		}
		if f.Range != nil {
			if f.Range.StartLine < 0 || f.Range.StartChar < 0 || f.Range.EndLine < 0 || f.Range.EndChar < 0 {
				return errors.New("voice transcript step: format range positions must be non-negative")
			}
		}
		return nil
	case "delete_file":
		if s.DeleteFileDirective == nil || s.EditDirective != nil || s.CommandDirective != nil || s.NavigationDirective != nil || s.UndoDirective != nil ||
			s.RenameDirective != nil || s.CodeActionDirective != nil || s.FormatDirective != nil ||
			s.MovePathDirective != nil || s.CreateFolderDirective != nil {
			return errors.New("voice transcript step: kind delete_file requires deleteFileDirective and no other directive payloads")
		}
		if strings.TrimSpace(s.DeleteFileDirective.Path) == "" {
			return errors.New("voice transcript step: delete_file requires path")
		}
		return nil
	case "move_path":
		if s.MovePathDirective == nil || s.EditDirective != nil || s.CommandDirective != nil || s.NavigationDirective != nil || s.UndoDirective != nil ||
			s.RenameDirective != nil || s.CodeActionDirective != nil || s.FormatDirective != nil ||
			s.DeleteFileDirective != nil || s.CreateFolderDirective != nil {
			return errors.New("voice transcript step: kind move_path requires movePathDirective and no other directive payloads")
		}
		if strings.TrimSpace(s.MovePathDirective.From) == "" || strings.TrimSpace(s.MovePathDirective.To) == "" {
			return errors.New("voice transcript step: move_path requires from and to")
		}
		return nil
	case "create_folder":
		if s.CreateFolderDirective == nil || s.EditDirective != nil || s.CommandDirective != nil || s.NavigationDirective != nil || s.UndoDirective != nil ||
			s.RenameDirective != nil || s.CodeActionDirective != nil || s.FormatDirective != nil ||
			s.DeleteFileDirective != nil || s.MovePathDirective != nil {
			return errors.New("voice transcript step: kind create_folder requires createFolderDirective and no other directive payloads")
		}
		if strings.TrimSpace(s.CreateFolderDirective.Path) == "" {
			return errors.New("voice transcript step: create_folder requires path")
		}
		return nil
	default:
		return fmt.Errorf("voice transcript step: unknown kind %q", s.Kind)
	}
}

func validateUndoDirective(u *UndoDirective) error {
	if u == nil {
		return errors.New("voice transcript step: undo requires undoDirective")
	}
	switch strings.TrimSpace(u.Scope) {
	case "last_edit", "last_transcript":
		return nil
	default:
		return fmt.Errorf("voice transcript step: undo scope must be last_edit or last_transcript, got %q", u.Scope)
	}
}

func validateNavigationDirective(n *NavigationDirective) error {
	if n == nil {
		return errors.New("voice transcript step: navigate requires navigationDirective")
	}
	switch n.Kind {
	case "success":
		if n.Action == nil {
			return errors.New("voice transcript step: navigate success requires navigationDirective.action")
		}
		if n.Reason != "" {
			return errors.New("voice transcript step: navigate success must not include reason")
		}
		return validateNavigationAction(n.Action)
	case "noop":
		if strings.TrimSpace(n.Reason) == "" {
			return errors.New("voice transcript step: navigate noop requires reason")
		}
		if n.Action != nil {
			return errors.New("voice transcript step: navigate noop must not include action")
		}
		return nil
	default:
		return fmt.Errorf("voice transcript step: unknown navigation directive kind %q", n.Kind)
	}
}

func validateNavigationAction(n *NavigationAction) error {
	if n == nil {
		return errors.New("voice transcript step: navigate requires navigationDirective.action")
	}
	kind := strings.TrimSpace(n.Kind)
	if kind == "" {
		return errors.New("voice transcript step: navigate requires non-empty navigationDirective.action.kind")
	}

	payloads := 0
	if n.OpenFile != nil {
		payloads++
	}
	if n.RevealSymbol != nil {
		payloads++
	}
	if n.MoveCursor != nil {
		payloads++
	}
	if n.SelectRange != nil {
		payloads++
	}
	if n.RevealEdit != nil {
		payloads++
	}
	if payloads != 1 {
		return errors.New("voice transcript step: navigate requires exactly one navigation action payload")
	}

	switch kind {
	case "open_file":
		if n.OpenFile == nil || strings.TrimSpace(n.OpenFile.Path) == "" {
			return errors.New("voice transcript step: open_file requires openFile.path")
		}
	case "reveal_symbol":
		if n.RevealSymbol == nil || strings.TrimSpace(n.RevealSymbol.SymbolName) == "" {
			return errors.New("voice transcript step: reveal_symbol requires revealSymbol.symbolName")
		}
	case "move_cursor":
		if n.MoveCursor == nil || n.MoveCursor.Target.Line < 0 || n.MoveCursor.Target.Char < 0 {
			return errors.New("voice transcript step: move_cursor requires moveCursor.target with non-negative line/char")
		}
	case "select_range":
		if n.SelectRange == nil {
			return errors.New("voice transcript step: select_range requires selectRange.target")
		}
		t := n.SelectRange.Target
		if t.StartLine < 0 || t.StartChar < 0 || t.EndLine < 0 || t.EndChar < 0 {
			return errors.New("voice transcript step: select_range requires non-negative target positions")
		}
	case "reveal_edit":
		if n.RevealEdit == nil || strings.TrimSpace(n.RevealEdit.EditId) == "" {
			return errors.New("voice transcript step: reveal_edit requires revealEdit.editId")
		}
	default:
		return fmt.Errorf("voice transcript step: unknown navigation kind %q", kind)
	}

	return nil
}

func (r VoiceTranscriptCompletion) Validate() error {
	if !r.Success {
		if r.Search != nil || r.Question != nil || r.Clarify != nil || r.FileSelection != nil || r.Workspace != nil {
			return errors.New("voice transcript result: grouped fields not allowed when success=false")
		}
		if strings.TrimSpace(r.UiDisposition) != "" {
			return errors.New("voice transcript result: uiDisposition must be omitted when success=false")
		}
	}
	if len([]rune(r.Summary)) > 8192 {
		return errors.New("voice transcript result: summary exceeds max length")
	}

	if r.Question != nil && r.Success {
		if strings.TrimSpace(r.Question.AnswerText) == "" {
			return errors.New("voice transcript result: question.answerText is required when question is set")
		}
		if len([]rune(r.Question.AnswerText)) > 8192 {
			return errors.New("voice transcript result: question.answerText exceeds max length")
		}
	}

	if r.Search != nil {
		if !r.Success {
			return errors.New("voice transcript result: search requires success=true")
		}
		s := r.Search
		for _, h := range s.Results {
			if strings.TrimSpace(h.Path) == "" {
				return errors.New("voice transcript result: search.results[].path is required")
			}
			if h.Line < 0 || h.Character < 0 {
				return errors.New("voice transcript result: search.results[] requires non-negative line/character")
			}
			if strings.Contains(h.Preview, "\u0000") {
				return errors.New("voice transcript result: search.results[].preview contains NUL")
			}
		}
		if s.ActiveIndex != nil {
			if *s.ActiveIndex < 0 {
				return errors.New("voice transcript result: search.activeIndex must be non-negative")
			}
		}
	}

	if r.Clarify != nil && r.Success {
		if strings.TrimSpace(r.Clarify.TargetResolution) == "" {
			return errors.New("voice transcript result: clarify.targetResolution is required when clarify is set")
		}
	}

	if r.FileSelection != nil && r.Success {
		fs := r.FileSelection
		for _, h := range fs.Results {
			if strings.TrimSpace(h.Path) == "" {
				return errors.New("voice transcript result: fileSelection.results[].path is required")
			}
			if strings.Contains(h.Preview, "\u0000") {
				return errors.New("voice transcript result: fileSelection.results[].preview contains NUL")
			}
		}
		n := len(fs.Results)
		if n > 0 {
			if fs.ActiveIndex == nil {
				return errors.New("voice transcript result: fileSelection.activeIndex is required when results is non-empty")
			}
			if *fs.ActiveIndex < 0 || int(*fs.ActiveIndex) >= n {
				return errors.New("voice transcript result: fileSelection.activeIndex out of range for results")
			}
		} else if fs.ActiveIndex != nil && *fs.ActiveIndex != 0 {
			return errors.New("voice transcript result: fileSelection.activeIndex must be omitted or 0 when results is empty")
		}
	}

	if r.Workspace != nil && r.Success {
		if !r.Workspace.NeedsFolder {
			return errors.New("voice transcript result: workspace object requires needsFolder=true")
		}
	}

	return nil
}
